hail/python/test/hailtop/test_dictfix.py

```diff
@@ -0,0 +1,87 @@
+from hailtop import dictfix
+
+
+def test_batch_example():
+    spec = {'input': dictfix.NoneOr({'logs': None}),
+            'main': dictfix.NoneOr({'logs': None}),
+            'output': dictfix.NoneOr({'logs': None})}
+
+    expected = {'input': None, 'main': None, 'output': None}
+
+    actual = dict()
+    dictfix.dictfix(actual, spec)
+    assert actual == expected
+
+    actual = {'input': None}
+    dictfix.dictfix(actual, spec)
+    assert actual == expected
+
+    actual = {'main': None}
+    dictfix.dictfix(actual, spec)
+    assert actual == expected
+
+    actual = {'main': None, 'output': None}
+    dictfix.dictfix(actual, spec)
+    assert actual == expected
+
+    expected = {'input': None,
+                'main': {'id': 3, 'logs': None},
+                'output': None}
+    actual = {'main': {'id': 3}, 'output': None}
+    dictfix.dictfix(actual, spec)
+    assert actual == expected
+
+    expected = {'input': None,
+                'main': {'id': 3, 'logs': 'abc\n123'},
+                'output': {'id': 4, 'logs': None}}
+    actual = {'main': {'id': 3, 'logs': 'abc\n123'}, 'output': {'id': 4}}
+    dictfix.dictfix(actual, spec)
+    assert actual == expected
+
+
+def test_type_asserts():
+    actual = {'x': 1}
+    try:
+        dictfix.dictfix(actual, {'x': str})
+    except AssertionError:
+        pass
+    else:
+        assert False, actual
+
+    actual = {'x': 'hi'}
```

Comment:
What happens if `actual = {'x': None}`? Does this raise an exception? Does the type parameter imply non-optional values?